### PR TITLE
feat: added File Upload Component

### PR DESCRIPTION
**Pull request by Swappea**
- Added new file upload components which is accessible in nature.
- Components are present to show list of uploaded files, show an image preview if needed, show file size, and ability to remove those files.

### CODE REVIEW

**Comment by Swappea** [comment]
@jer3m01 - I have fixed the formatting issues, could you please approve the workflow so I can check if formatting is as expected?

Also could you please review the PR?

`.mdx` formatting is very weird in terms of DX.

**Comment by jer3m01** [comment]
> `.mdx` formatting is very weird in terms of DX.

Yeah the docs aren't the nicest to write, this is being worked on 👀.

**Comment by Swappea** [comment]
> > `.mdx` formatting is very weird in terms of DX.
> 
> Yeah the docs aren't the nicest to write, this is being worked on 👀.

@jer3m01 - any idea on how to fix the formatting issue I am getting? Within VSCode, if I format using prettier I don't see any issues.

Is there any command that I am probably missing?

**Comment by jer3m01** [comment]
> Is there any command that I am probably missing?

You can run `pnpm -w format` to run both formatters.

File: packages/core/src/file-upload/file-upload-hidden-input.tsx [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+) {
+	const context = useFileUploadContext();
+
+	const onInput: JSX.EventHandler<HTMLInputElement, InputEvent> = (event) => {
+		if (context.disabled) {
+			return;
+		}
+
+		const { files } = event.currentTarget;
+		console.log("files", files);
```
Comment (jer3m01) [inline packages/core/src/file-upload/file-upload-hidden-input.tsx:31]: Leftover console.log

File: packages/core/src/file-upload/file-upload-hidden-input.tsx [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+		const t = Array.from(files ?? []);
```
Comment (jer3m01) [inline packages/core/src/file-upload/file-upload-hidden-input.tsx:32]: `t` seems unused?

File: packages/core/src/file-upload/file-upload-hidden-input.tsx [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+	return (
+		<Polymorphic
+			as="input"
+			type="file"
+			id={context.inputId}
+			accept={context.accept}
+			multiple={context.multiple}
+			ref={(el: HTMLInputElement) => (context.fileInputRef = el)}
+			style={{ display: "none" }}
```
Comment (jer3m01) [inline packages/core/src/file-upload/file-upload-hidden-input.tsx:44]: For hiding an input please use `visuallyHiddenStyles` from `@kobalte/utils` ([example here](https://github.com/kobaltedev/kobalte/blob/76319603b4b5242c2533f1cc5e910a3d9cd51d55/packages/core/src/switch/switch-input.tsx#L140)).

File: packages/core/src/file-upload/file-upload-item-delete-trigger.tsx [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+		if (context.disabled) {
+			return;
+		}
+		context.removeFile(file);
+	};
+
+	return (
+		<Polymorphic
+			as="button"
+			class="file-upload__item-delete-trigger"
```
Comment (jer3m01) [inline packages/core/src/file-upload/file-upload-item-delete-trigger.tsx:39]: Classes shouldn't be applied by default.

File: packages/core/src/file-upload/file-upload-item-group.tsx [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+export type FileUploadItemGroupRootProps<
+	T extends ValidComponent | HTMLElement = HTMLElement,
+> = Partial<FileUploadItemGroupCommonProps<ElementOf<T>>>;
+
+export function FileUploadItemGroup<T extends ValidComponent = "ul">(
+	props: PolymorphicProps<T, FileUploadItemGroupRootProps<T>>,
+) {
+	return (
+		<Polymorphic as="ul" class="file-upload__item-group" {...props}>
```
Comment (jer3m01) [inline packages/core/src/file-upload/file-upload-item-group.tsx:23]: Classes shouldn't be applied by default.

**Comment by Swappea** [comment]
> > Is there any command that I am probably missing?
> 
> You can run `pnpm -w format` to run both formatters.

Wow for some reason after running this command, over 500 files were changed as per git but there were no visible changes.
Need to check what happened.

I have pushed the 2 files which were showing issues.

**Review by jer3m01 — CHANGES_REQUESTED** [review]
Thanks for the interest and the work!

I've noted a few issues but it mostly looks good.

Could another example be added to the docs for HTML form submissions, like there is for [`TextField`](https://deploy-preview-506--kobalte.netlify.app/docs/core/components/text-field#html-forms) but instead maybe only write the submitted file names in the `alert()`.

Thanks again for the PR!

File: packages/core/src/file-upload/file-upload-hidden-input.tsx (hunk 1 above)
Comment (Swappea) [reply packages/core/src/file-upload/file-upload-hidden-input.tsx]: Removed

File: packages/core/src/file-upload/file-upload-dropzone.tsx [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+export function FileUploadDropZone<T extends ValidComponent = "div">(
+	props: PolymorphicProps<T, FileUploadDropZoneRootProps<T>>,
+) {
+	const [isDragging, setIsDragging] = createSignal(false);
+	const context = useFileUploadContext();
+
+	return (
+		<Polymorphic
+			as="div"
+			class="file-upload__dropzone"
```
[the remark this answers is not in the document]
Comment (Swappea) [reply packages/core/src/file-upload/file-upload-dropzone.tsx]: Done

File: packages/core/src/file-upload/file-upload-hidden-input.tsx (hunk 3 above)
Comment (Swappea) [reply packages/core/src/file-upload/file-upload-hidden-input.tsx]: Done

File: packages/core/src/file-upload/file-upload-item-group.tsx (hunk 5 above)
Comment (Swappea) [reply packages/core/src/file-upload/file-upload-item-group.tsx:23]: Removed

File: packages/core/src/file-upload/file-upload-item-preview-image.tsx [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+		}
+
+		const cleanup = createFileUrl(file, (url) => setUrl(url));
+		onCleanup(cleanup);
+	});
+
+	return (
+		<Polymorphic
+			as="img"
+			class="file-upload__item-preview-image"
```
[the remark this answers is not in the document]
Comment (Swappea) [reply packages/core/src/file-upload/file-upload-item-preview-image.tsx:53]: Removed

File: packages/core/src/file-upload/file-upload-item.tsx [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+export type FileUploadItemRootProps<
+	T extends ValidComponent | HTMLElement = HTMLElement,
+> = FileUploadItemOptions & Partial<FileUploadItemCommonProps<ElementOf<T>>>;
+
+export function FileUploadItem<T extends ValidComponent = "li">(
+	props: PolymorphicProps<T, FileUploadItemRootProps<T>>,
+) {
+	return (
+		<FileUploadItemProvider file={props.file}>
+			<Polymorphic as="li" class="file-upload__item" {...props}>
```
[the remark this answers is not in the document]
Comment (Swappea) [reply packages/core/src/file-upload/file-upload-item.tsx]: removed

File: packages/core/src/file-upload/file-upload-root.tsx [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+export type FileUploadRootProps<
+	T extends ValidComponent | HTMLElement = HTMLElement,
+> = FileUploadRootOptions & Partial<FileUploadCommonProps<ElementOf<T>>>;
+
+export function FileUpload<T extends ValidComponent = "div">(
+	props: PolymorphicProps<T, FileUploadRootProps<T>>,
+) {
+	return (
+		<FileUploadProvider {...props}>
+			<Polymorphic as="div" class="file-upload__root" {...props}>
```
[the remark this answers is not in the document]
Comment (Swappea) [reply packages/core/src/file-upload/file-upload-root.tsx]: removed

File: packages/core/src/file-upload/file-upload-trigger.tsx [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+export function FileUploadTrigger<T extends ValidComponent = "input">(
+	props: PolymorphicProps<T, FileUploadTriggerRootProps<T>>,
+) {
+	const context = useFileUploadContext();
+
+	return (
+		<Polymorphic
+			disabled={!!context.disabled}
+			as="button"
```
[the remark this answers is not in the document]
Comment (Swappea) [reply packages/core/src/file-upload/file-upload-trigger.tsx]: done

File: packages/core/src/file-upload/file-upload-trigger.tsx [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+			type="button"
+			class="file-upload__trigger"
```
[the remark this answers is not in the document]
Comment (Swappea) [reply packages/core/src/file-upload/file-upload-trigger.tsx]: removed

File: packages/core/src/file-upload/file-upload-label.tsx [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+
+export function FileUploadLabel<T extends ValidComponent = "label">(
+	props: PolymorphicProps<T, FileUploadLabelRootProps<T>>,
+) {
+	const context = useFileUploadContext();
+
+	return (
+		<Polymorphic
+			as="label"
+			class="file-upload__label"
```
[the remark this answers is not in the document]
Comment (Swappea) [reply packages/core/src/file-upload/file-upload-label.tsx]: removed

File: packages/core/src/file-upload/file-upload-item-size.tsx [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+	return `${Number.parseFloat((bytes / k ** i).toFixed(precision))} ${sizes[i]}`;
+}
+
+export function FileUploadItemSize<T extends ValidComponent = "div">(
+	props: PolymorphicProps<T, FileUploadItemSizeRootProps<T>>,
+) {
+	const { file } = useFileUploadItemContext();
+
+	return (
+		<Polymorphic as="div" class="file-upload__item-size" {...props}>
```
[the remark this answers is not in the document]
Comment (Swappea) [reply packages/core/src/file-upload/file-upload-item-size.tsx:40]: removed

File: packages/core/src/file-upload/file-upload-item-preview.tsx [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+	props: PolymorphicProps<T, FileUploadItemPreviewRootProps<T>>,
+) {
+	const { file } = useFileUploadItemContext();
+
+	if (!file.type.match(props.type ?? ".*")) {
+		return null;
+	}
+
+	return (
+		<Polymorphic as="div" class="file-upload__item-preview" {...props}>
```
[the remark this answers is not in the document]
Comment (Swappea) [reply packages/core/src/file-upload/file-upload-item-preview.tsx]: removed

File: packages/core/src/file-upload/file-upload-item-preview.tsx [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+	T extends ValidComponent | HTMLElement = HTMLElement,
+> = FileUploadItemPreviewOptions &
+	Partial<FileUploadItemPreviewCommonProps<ElementOf<T>>>;
+
+export function FileUploadItemPreview<T extends ValidComponent = "div">(
+	props: PolymorphicProps<T, FileUploadItemPreviewRootProps<T>>,
+) {
+	const { file } = useFileUploadItemContext();
+
+	if (!file.type.match(props.type ?? ".*")) {
```
[the remark this answers is not in the document]
Comment (Swappea) [reply packages/core/src/file-upload/file-upload-item-preview.tsx]: habits of being a react developer.. fixed

File: packages/core/src/file-upload/file-upload-hidden-input.tsx (hunk 2 above)
Comment (Swappea) [reply packages/core/src/file-upload/file-upload-hidden-input.tsx]: removed

**Comment by Swappea** [comment]
@jer3m01 - updated please check..
Thank you

**Comment by jer3m01** [comment]
Thanks for the changes! A few classes haven't been removed yet, I unresolved the cases.

**Comment by Swappea** [comment]
Sorry, missed removing those, not sure how. All the classes have not been removed.

**Review by jer3m01 — APPROVED** [review]
Looks good! Thanks for the PR and changes!

Will merge soon.

**Pull request merged**